pkg/chain/chain.go

```diff
@@ -1,27 +1,83 @@
-// Package eth contains interface for interaction with an ethereum blockchain
-// along with structures reflecting events emitted on an ethereum blockchain.
-package eth // TODO: rename; this can be any host chain
+// Package chain contains interface for interaction with a host chain along with
+// structures reflecting events emitted on such a chain.
+package chain // TODO: rename; this can be any host chain
 
 import (
+	cecdsa "crypto/ecdsa"
+	"fmt"
 	"math/big"
 	"time"
 
-	"github.com/ethereum/go-ethereum/common"
 	"github.com/keep-network/keep-common/pkg/subscription"
 	"github.com/keep-network/keep-core/pkg/chain"
 	"github.com/keep-network/keep-ecdsa/pkg/ecdsa"
 )
 
-// Handle represents a handle to an ethereum blockchain.
+type KeepApplicationID interface {
+	fmt.Stringer
+}
+
+type KeepMemberID interface {
+	fmt.Stringer
+
+	// KeepMemberIDs should be convertible to their associated OperatorID. This
+	// may be a 1-to-1 correspondence, i.e. a simple interface cast, or it may
+	// be an internally-tracked association on the host chain.
+	OperatorID() OperatorID
+}
+
+type OperatorID interface {
+	fmt.Stringer
+
+	// OperatorIDs should be convertible to an associated KeepMemberID given the
+	// id of the keep in question. This may be a 1-to-1 correspondence, i.e. a
+	// simple interface cast, or it may be an internally-tracked association on
+	// the host chain. The conversion _may but is not guaranteed to be_ nil if
+	// the operator is not in the specified keep.
+	KeepMemberID(keepID KeepID) KeepMemberID
+}
+
+// Should be chain-specific, unique across chains, and simple ([A-Za-z0-9-_.]).
+// This can generally mean prefixing the common name of the chain followed by a
+// dash followed by the chain-specific identifier (e.g., the contract address
+// for Ethereum). For legacy reasons, Ethereum KeepIDs should always be the bare
+// Keep contract address.
+//
+// Should not be expected to be comparable between types or amongst themselves.
+type KeepID interface {
+	fmt.Stringer
+}
+
+// package chain in keep-core -- or keep-common/chain?
+// type BaseHandle interface {
+// 	BlockCounter() (BlockCounter, error)
+// 	StakeMonitor() (StakeMonitor, error)
+// 	BalanceMonitor() (BalanceMonitor, error)
+// }
+
+// FIXME BalanceMonitor may be an Ethereum-specific concept, kick it
+// FIXME from the chain interface?
```

Comment:
If this is more than 10 minutes of work, I'd consider leaving FIXME but not bothering about it for now. All the potential integrations I can think of have a concept of operator's account balance.

Reply:
Yeah, I don't think this one's that important. I think BlockCounter is going to have to be reworked though.